packages/engine/src/grabbable/GrabbableComponent.ts

```diff
@@ -0,0 +1,170 @@
+import { InputSourceComponent } from '@ir-engine/spatial/src/input/components/InputSourceComponent'
+import { InputState } from '@ir-engine/spatial/src/input/state/InputState'
+
+import { S } from '@ir-engine/ecs/src/schemas/JSONSchemas'
+import { NetworkTopics } from '@ir-engine/network'
+import { AvatarComponent } from '../avatar/components/AvatarComponent'
+import { InteractableComponent, XRUIVisibilityOverride } from '../interaction/components/InteractableComponent'
+
+// @todo move this to spatial package schema definitions
+export const XRHandedness = S.LiteralUnion(['none', 'left', 'right'], 'none')
+
+/**
+ * GrabbableComponent
+ * - Allows an entity to be grabbed by an entity with a GrabberComponent
+ */
+export const GrabbableComponent = defineComponent({
+  name: 'GrabbableComponent',
+  jsonID: 'EE_grabbable',
+
+  toJSON: () => true,
+
+  grabbableCallbackName: 'grabCallback',
+
+  reactor: function () {
+    const entity = useEntityContext()
+    const isGrabbed = !!useOptionalComponent(entity, GrabbedComponent)
+    const interactableComponent = useComponent(entity, InteractableComponent)
+
+    useEffect(() => {
+      if (isClient) {
+        setCallback(entity, GrabbableComponent.grabbableCallbackName, (entity: Entity, targetEntity: Entity) =>
+          grabCallback(entity)
+        )
+      }
+    }, [])
+
+    useEffect(() => {
+      interactableComponent.uiVisibilityOverride.set(
+        isGrabbed ? XRUIVisibilityOverride.off : XRUIVisibilityOverride.none
+      )
+    }, [isGrabbed, !!interactableComponent])
+
+    return null
+  },
+
+  grab: (grabberEntity: Entity, grabbableEntity: Entity, handedness = getState(InputState).preferredHand) => {
+    if (
+      !grabberEntity ||
+      !hasComponent(grabberEntity, GrabberComponent) ||
+      !hasComponent(grabbableEntity, GrabbableComponent)
+    )
+      return
+
+    const grabber = getComponent(grabberEntity, GrabberComponent)
+    const grabbedEntity = grabber[handedness]!
+    if (grabbedEntity) return
+    dispatchAction(
+      GrabbableNetworkAction.setGrabbedObject({
+        entityUUID: getComponent(grabbableEntity, UUIDComponent),
+        grabberEntityUUID: getComponent(grabberEntity, UUIDComponent),
+        grabbed: true,
+        attachmentPoint: handedness
+      })
+    )
+  },
+
+  drop: (grabberEntity: Entity, grabbableEntity: Entity) => {
+    if (
+      !grabberEntity ||
+      !hasComponent(grabberEntity, GrabberComponent) ||
+      !hasComponent(grabbableEntity, GrabbableComponent)
+    )
+      return
+
+    dispatchAction(
+      GrabbableNetworkAction.setGrabbedObject({
+        entityUUID: getComponent(grabbableEntity, UUIDComponent),
+        grabberEntityUUID: getComponent(grabberEntity, UUIDComponent),
+        grabbed: false
+      })
+    )
+  }
+})
+
+const grabCallback = (grabbableEntity: Entity) => {
+  const nonCapturedInputSources = InputSourceComponent.nonCapturedInputSources()
```

Comment:
Grabbing all non-captured input sources here probably isn't the behavior we need. I think we want to know which input source triggered the grab callback, and use that to determine which hand the grabbable should be attached to. We can pass the input source down from the InteractComponent's callback.

Reply:
that information does not come from InteractableComponent currently. how do you suggest we do this? might be out of scope for the context of this PR